frontend/test/pytest/test_callback.py

```diff
@@ -586,5 +590,696 @@ def wrapper():
+        return ({"x": vjp0, "y": vjp1},)
+
+    @qml.qjit
+    @grad
+    def result(weights):
+        weights["y"] = weights["y"] ** 2
+        res = some_func(weights)
+        return jnp.dot(res["one"], res["two"])
+
+    @jax.jit
+    @jax.grad
+    def expected(weights):
+        x = weights["x"]
+        y = weights["y"]
+        return jnp.dot(jnp.sin(x) @ y**2, y**4)
+
+    weights = {"x": jnp.array([[1.0, 0.5], [0.12, -1.2]]), "y": jnp.array([-0.6, 0.2])}
+    flat_results_obs, obs_shape = jax._src.tree_util.tree_flatten(result(weights))
+    flat_results_exp, exp_shape = jax._src.tree_util.tree_flatten(expected(weights))
+    assert obs_shape == exp_shape
+    for obs, exp in zip(flat_results_obs, flat_results_exp):
+        assert np.allclose(obs, exp)
+
+    # {'x': Array([[2.5208345e-02, 4.5493888e-03],
+    #        [5.7185785e-04, 2.3190898e-05]], dtype=float64),
+    # 'y': Array([-0.40939555,  0.02444299], dtype=float64)}
+
+
+def test_callback_backwards_function():
+    """Test a workflow where the bwd function is also using
+    a callback to compute the VJP"""
+
+    shape = {
+        "one": jax.ShapeDtypeStruct((2,), jnp.float64),
+        "two": jax.ShapeDtypeStruct((2,), jnp.float64),
+    }
+    vjp_shape = {
+        "x": jax.ShapeDtypeStruct((2, 2), jnp.float64),
+        "y": jax.ShapeDtypeStruct((2,), jnp.float64),
+    }
+
+    @pure_callback
+    def some_func(weights) -> shape:
+        return {"one": np.sin(weights["x"]) @ weights["y"], "two": weights["y"] ** 2}
+
+    @some_func.fwd
+    def some_func_fwd(weights):
+        return some_func(weights), weights
+
+    @pure_callback
+    def some_func_bwd_vjp(res, dy) -> vjp_shape:
+        vjp0 = res["y"] * np.cos(res["x"]) * np.reshape(dy["one"], (-1, 1))
+        vjp1 = dy["one"] @ np.sin(res["x"]) + 2 * res["y"] * dy["two"]
+        return ({"x": vjp0, "y": vjp1},)
+
+    @some_func.bwd
+    def some_func_bwd(res, dy):
+        return some_func_bwd_vjp(res, dy)
+
+    @qml.qjit
+    @grad
+    def result(weights):
+        weights["y"] = weights["y"] ** 2
+        res = some_func(weights)
+        return jnp.dot(res["one"], res["two"])
+
+    @jax.jit
+    @jax.grad
+    def expected(weights):
+        x = weights["x"]
+        y = weights["y"]
+        return jnp.dot(jnp.sin(x) @ y**2, y**4)
+
+    weights = {"x": jnp.array([[1.0, 0.5], [0.12, -1.2]]), "y": jnp.array([-0.6, 0.2])}
+    flat_results_obs, obs_shape = jax._src.tree_util.tree_flatten(result(weights))
+    flat_results_exp, exp_shape = jax._src.tree_util.tree_flatten(expected(weights))
+    assert obs_shape == exp_shape
+    for obs, exp in zip(flat_results_obs, flat_results_exp):
+        assert np.allclose(obs, exp)
+
+
+def test_different_shapes():
+    """Different input output shape"""
+
+    def fun(x):
+        y = jnp.array([3.0, 4.0, 5.0])
+        return y * x[0] + x[1]
+
+    # hack for the type
+    ty = jnp.array([2.0, 2.0, 2.0])
+
+    @pure_callback
+    def fun_callback(x) -> ty:
```

Comment:
@josh146 this makes me think we really need a simpler of specifying array types 😅

Reply:
Jax suffers from the same issue :thinking: Here, though, I always know that a value can be mapped to a type, so I just make values because I don't like making ShapedArray or the other version that Josh uses :sweat_smile:

Reply:
Absolutely, that doesn't mean we couldn't improve it though :)

Reply:
For comparison, the ShapedArray way `jax.core.ShapedArray((3,), float)`. It does feel annoying to type, I agree, although I'm not sure why, it's only slightly longer in this case. 

And it does seem harder to do better than `fun(shape, dtype)` 🤔